docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,247 @@
+  * Removes the finalizer from the `ObjectMeta.Finalizers` of the parent object.
+
+## Expected race conditions
+
+* *Controller of the parent resource creates a child object after the orphaning finalizer has updated the ParentReferences of other children*: This can be solved by implementing Status.ObservedGeneration in all resources, so that before applying the orphaning logic to children, the orphaning finalizer checks Status.ObservedGeneration to ensure the controller of the parent resource has observed the deletion. Note that in design #1, the same [race condition](#expected-race-conditions) is solved by introducing the tombstone.
+
+# Comparison between design #1 and #2
+
+* Advantages of "Tombstone + GC" design
+  * Faster to free the resource name compared to using finalizers. The original object can be deleted to free the resource name once the tombstone is created, rather than waiting for the finalizers to update all children’s ObejectMeta.ParentReferences.
+  * Doesn't require implementing ObservedGeneration in all resources.
+* Advantages of "Finalizer Framework + GC"
+  * The finalizer framework is needed for other purposes as well.
+
+# Common designs
+
+## Orphan adoption
+
+Controllers are responsible for adopting orphaned child resources. To do so, controllers
+* Checks a potential child object’s ParentReferences to determine if it is orphaned.
+* Fills the ParentReferences if the object matches the controller’s selector and is orphaned.
+
+There is a potential race between the orphaning finalizer and the controllers, where the orphaning finalizers removes the controller from the child object's ParentReference list, and the controller adds itself back. This can be solved by the [ObservedGeneration mechanism](#expected-race-conditions-1) described previously.
+
+## Upgrading a cluster to support cascading deletion
+
+For the master, after upgrading to a version that supports cascading deletion, the ParentReferences of existing objects remain empty, so the controllers will regard them as orphaned and start the adoption procedures. After the adoptions are done, server-side cascading will be effective for these existing objects.
+
+For nodes, cascading deletion does not affect them.
+
+For kubectl, we will keep the kubectl’s cascading deletion logic for one more release.
+
+## Recovering from abnormal cascading deletion
+
+In case the garbage collector is mistakenly deleting objects, we should provide mechanism to stop the garbage collector and restore the objects.
+
+* Stopping the garbage collector
+
+  We will add a "--enable-garbage-collector" flag to the controller manager binary to indicate if the garbage collector should be enabled. Admin can stop the garbage collector in a running cluster by restarting the kube-conntroller-manager with --enable-garbe-collector=false.
+
+* Restoring mistakenly deleted objects
```

Comment:
We can build more automation in to the restoration mechanism, like restoring objects deleted after a certain timestamp.

Also it would be handy If we can record who sends the deletion request.